File: PnmReactNativeDemo/src/services/LiveKitService.ts
import {Room, RoomEvent, ConnectionState, Track, LocalTrackPublication} from 'livekit-client';
import type {E2EEOptions, RoomOptions} from 'livekit-client';
import {RNKeyProvider, RNE2EEManager} from '@livekit/react-native';

/**
 * Callbacks for native media state changes.
 */
export interface LiveKitServiceCallbacks {
  onConnected?: () => void;
  onDisconnected?: (reason?: string) => void;
  onError?: (error: Error) => void;
  onTrackPublished?: (source: 'mic' | 'webcam' | 'screenshare') => void;
  onTrackUnpublished?: (source: 'mic' | 'webcam' | 'screenshare') => void;
  onMuted?: (source: 'mic' | 'webcam' | 'screenshare', muted: boolean) => void;
  onScreenShareRequest?: () => void;
}

/**
 * Manages a native LiveKit publisher connection.
 * Connects with the native twin token, publishes mic/webcam on demand.
 */
export class LiveKitService {
  private room: Room | null = null;
  private callbacks: LiveKitServiceCallbacks = {};
  private micPublication: LocalTrackPublication | undefined;
  private webcamPublication: LocalTrackPublication | undefined;
  private screenSharePublication: LocalTrackPublication | undefined;
  private keyProvider: RNKeyProvider | null = null;
  private e2eeManager: RNE2EEManager | null = null;

  setCallbacks(cbs: LiveKitServiceCallbacks): void {
    this.callbacks = cbs;
  }

  /**
   * Connect to LiveKit with the native twin token.
   * If already connected, tears down the previous connection first.
   */
  async connect(url: string, token: string, e2eeKey?: string): Promise<void> {
    // Re-initialization: tear down existing connection first
    if (this.room) {
      // Clean up old E2EE before disconnect
      if (this.e2eeManager) {
        this.e2eeManager = null;
      }
      if (this.keyProvider) {
        this.keyProvider.dispose();
        this.keyProvider = null;
      }
      await this.disconnect();
    }

    // Prepare room options
    const roomOptions: RoomOptions = {
      publishDefaults: {
        videoCodec: 'h264',
      },
    };

    // Set up E2EE if key is provided
    if (e2eeKey) {
      // Clean up any previous E2EE instances
      this.keyProvider?.dispose();
      this.e2eeManager = null;
      this.keyProvider = null;

      this.keyProvider = new RNKeyProvider({ sharedKey: true });
      await this.keyProvider.setSharedKey(e2eeKey);
      this.e2eeManager = new RNE2EEManager(this.keyProvider);

      roomOptions.e2ee = {
        keyProvider: this.keyProvider,
        e2eeManager: this.e2eeManager,
      } as E2EEOptions;
    }

    this.room = new Room(roomOptions);

    if (this.e2eeManager) {
      this.e2eeManager.setup(this.room);
    }

    this.room.on(RoomEvent.Disconnected, reason => {
      this.callbacks.onDisconnected?.(reason?.toString());
    });

    this.room.on(RoomEvent.ConnectionStateChanged, (state: ConnectionState) => {
      if (state === ConnectionState.Connected) {
        this.callbacks.onConnected?.();
      }
    });

    try {
      await this.room.connect(url, token, {
        autoSubscribe: false, // as we've only publisher permission
      });
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Publish microphone */
  async enableMic(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      this.micPublication = await this.room.localParticipant.setMicrophoneEnabled(
        true,
      );
      this.callbacks.onTrackPublished?.('mic');
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Unpublish microphone */
  async disableMic(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      const publication = this.room.localParticipant.getTrackPublication(Track.Source.Microphone);
      if (publication && this.micPublication) { // Ensure we have a publication to unpublish and a local reference
        await this.room.localParticipant.unpublishTrack(this.micPublication.track);
        this.micPublication = undefined;
      }
      this.callbacks.onTrackUnpublished?.('mic');
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Publish webcam */
  async enableWebcam(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      this.webcamPublication = await this.room.localParticipant.setCameraEnabled(
        true,
      );
      this.callbacks.onTrackPublished?.('webcam');
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Unpublish webcam */
  async disableWebcam(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      const publication = this.room.localParticipant.getTrackPublication(Track.Source.Camera);
      if (publication && this.webcamPublication) { // Ensure we have a publication to unpublish and a local reference
        await this.room.localParticipant.unpublishTrack(this.webcamPublication.track);
        this.webcamPublication = undefined;
      }
      this.callbacks.onTrackUnpublished?.('webcam');
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Mute microphone (keep track published, no audio) */
  async muteMic(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      await this.room.localParticipant.setMicrophoneEnabled(false);
      this.callbacks.onMuted?.('mic', true);
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Unmute microphone */
  async unmuteMic(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      await this.room.localParticipant.setMicrophoneEnabled(true);
      this.callbacks.onMuted?.('mic', false);
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Mute webcam (keep track published, no video) */
  async muteWebcam(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      await this.room.localParticipant.setCameraEnabled(false);
      this.callbacks.onMuted?.('webcam', true);
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Unmute webcam */
  async unmuteWebcam(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      await this.room.localParticipant.setCameraEnabled(true);
      this.callbacks.onMuted?.('webcam', false);
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Publish screen share */
  async enableScreenShare(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      const existing = this.room.localParticipant.getTrackPublication(Track.Source.ScreenShare);
      if (existing && this.screenSharePublication) {
        return; // duplicate guard
      }
      this.screenSharePublication = await this.room.localParticipant.setScreenShareEnabled(true);
      this.callbacks.onTrackPublished?.('screenshare');
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Unpublish screen share */
  async disableScreenShare(): Promise<void> {
    if (!this.room) {
      return;
    }
    try {
      await this.room.localParticipant.setScreenShareEnabled(false);
      this.screenSharePublication = undefined;
      this.callbacks.onTrackUnpublished?.('screenshare');
    } catch (err) {
      this.callbacks.onError?.(err as Error);
      throw err;
    }
  }

  /** Disconnect from LiveKit and release all media */
  async disconnect(): Promise<void> {
    if (this.room) {
      await this.room.disconnect(true);
      this.room = null;
    }
    if (this.e2eeManager) {
      this.e2eeManager = null;
    }
    if (this.keyProvider) {
      this.keyProvider.dispose();
      this.keyProvider = null;
    }
    this.micPublication = undefined;
    this.webcamPublication = undefined;
    this.screenSharePublication = undefined;
  }
}

File: PnmNativeAndroidDemo/app/src/main/java/com/plugnmeet/nativedemo/services/LiveKitService.kt
package com.plugnmeet.nativedemo.services

import android.content.Context
import android.content.Intent
import android.util.Log
import io.livekit.android.LiveKit
import io.livekit.android.LiveKitOverrides
import io.livekit.android.AudioOptions
import io.livekit.android.RoomOptions
import io.livekit.android.e2ee.E2EEOptions
import io.livekit.android.events.RoomEvent
import io.livekit.android.events.collect
import io.livekit.android.room.Room
import io.livekit.android.room.participant.VideoTrackPublishDefaults
import io.livekit.android.room.track.Track
import io.livekit.android.room.track.screencapture.ScreenCaptureParams
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Manages a native LiveKit publisher connection.
 * Connects with the native twin token, publishes mic/webcam on demand.
 * Follows the LiveKit Android SDK patterns from the sample app.
 */
class LiveKitService {

    companion object {
        private const val TAG = "LiveKitService"
    }

    interface Callbacks {
        fun onConnected()
        fun onDisconnected(reason: String? = null)
        fun onError(error: Exception)
        fun onTrackPublished(source: String) // "mic" or "webcam"
        fun onTrackUnpublished(source: String)
        fun onMuted(source: String, muted: Boolean)
    }

    private var room: Room? = null
    private var callbacks: Callbacks? = null

    fun getRoom(): Room? = room
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    fun setCallbacks(cbs: Callbacks) {
        callbacks = cbs
    }

    /**
     * Connect to LiveKit with the native twin token.
     * If already connected, tears down the previous connection first.
     *
     * @param e2eeKey optional shared E2EE key. When non-null, end-to-end encryption
     * is enabled for the published tracks using the same shared key as the web client.
     */
    suspend fun connect(context: Context, url: String, token: String, e2eeKey: String? = null) {
        if (room != null) {
            disconnect()
        }

        try {
            // Shared-key E2EE (matches the web client's ExternalE2EEKeyProvider shared key).
            // The Room creates/owns the E2EEManager when e2eeOptions is set.
            val e2eeOptions = if (!e2eeKey.isNullOrEmpty()) {
                Log.d(TAG, "E2EE enabled for native publisher")
                E2EEOptions(sharedKey = e2eeKey)
            } else {
                null
            }

            // H.264 for hardware encoder support on Android (lower CPU/battery vs VP8).
            // Let the SDK auto-select bitrate/FPS based on source and network conditions.
            val lkRoom = LiveKit.create(
                appContext = context,
                options = RoomOptions(
                    e2eeOptions = e2eeOptions,
                    videoTrackPublishDefaults = VideoTrackPublishDefaults(
                        videoCodec = "h264"
                    )
                ),
                overrides = LiveKitOverrides(
                    audioOptions = AudioOptions()
                )
            )

            room = lkRoom
            lkRoom.connect(url = url, token = token)

            // Observe room events for connection state and track changes
            scope.launch {
                lkRoom.events.collect { event ->
                    when (event) {
                        is RoomEvent.Connected -> {
                            Log.d(TAG, "Connected to LiveKit")
                            callbacks?.onConnected()
                        }
                        is RoomEvent.Disconnected -> {
                            Log.d(TAG, "Disconnected: ${event.reason}")
                            callbacks?.onDisconnected(event.reason.toString())
                        }
                        is RoomEvent.Reconnecting -> {
                            Log.d(TAG, "Reconnecting...")
                        }
                        is RoomEvent.FailedToConnect -> {
                            Log.e(TAG, "Failed to connect", event.error)
                            callbacks?.onError(Exception(event.error.message))
                        }
                        is RoomEvent.TrackPublished -> {
                            if (event.participant is io.livekit.android.room.participant.LocalParticipant) {
                                when (event.publication.source) {
                                    Track.Source.MICROPHONE -> callbacks?.onTrackPublished("mic")
                                    Track.Source.CAMERA -> callbacks?.onTrackPublished("webcam")
                                    Track.Source.SCREEN_SHARE -> callbacks?.onTrackPublished("screenshare")
                                    else -> Unit
                                }
                            }
                        }
                        is RoomEvent.TrackUnpublished -> {
                            if (event.participant is io.livekit.android.room.participant.LocalParticipant) {
                                when (event.publication.source) {
                                    Track.Source.MICROPHONE -> callbacks?.onTrackUnpublished("mic")
                                    Track.Source.CAMERA -> callbacks?.onTrackUnpublished("webcam")
                                    Track.Source.SCREEN_SHARE -> callbacks?.onTrackUnpublished("screenshare")
                                    else -> Unit
                                }
                            }
                        }
                        is RoomEvent.TrackMuted -> {
                            if (event.participant is io.livekit.android.room.participant.LocalParticipant) {
                                when (event.publication.source) {
                                    Track.Source.MICROPHONE -> callbacks?.onMuted("mic", true)
                                    Track.Source.CAMERA -> callbacks?.onMuted("webcam", true)
                                    else -> Unit
                                }
                            }
                        }
                        is RoomEvent.TrackUnmuted -> {
                            if (event.participant is io.livekit.android.room.participant.LocalParticipant) {
                                when (event.publication.source) {
                                    Track.Source.MICROPHONE -> callbacks?.onMuted("mic", false)
                                    Track.Source.CAMERA -> callbacks?.onMuted("webcam", false)
                                    else -> Unit
                                }
                            }
                        }
                        else -> Unit
                    }
                }
            }

        } catch (e: Exception) {
            Log.e(TAG, "Failed to connect", e)
            callbacks?.onError(e)
            throw e
        }
    }

    /** Publish microphone */
    suspend fun enableMic() {
        val room = room ?: return
        try {
            room.localParticipant.setMicrophoneEnabled(true)
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Unpublish microphone */
    suspend fun disableMic() {
        val room = room ?: return
        try {
            val pub = room.localParticipant.getTrackPublication(Track.Source.MICROPHONE)
            pub?.track?.let { room.localParticipant.unpublishTrack(it) }
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Publish webcam */
    suspend fun enableWebcam() {
        val room = room ?: return
        try {
            room.localParticipant.setCameraEnabled(true)
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Unpublish webcam */
    suspend fun disableWebcam() {
        val room = room ?: return
        try {
            val pub = room.localParticipant.getTrackPublication(Track.Source.CAMERA)
            pub?.track?.let { room.localParticipant.unpublishTrack(it) }
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Publish screen share */
    suspend fun enableScreenShare(data: Intent) {
        val room = room ?: return
        if (room.localParticipant.getTrackPublication(Track.Source.SCREEN_SHARE) != null) {
            Log.d(TAG, "Screen share already published")
            return
        }
        try {
            room.localParticipant.setScreenShareEnabled(true, ScreenCaptureParams(data))
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Unpublish screen share */
    suspend fun disableScreenShare() {
        val room = room ?: return
        if (room.localParticipant.getTrackPublication(Track.Source.SCREEN_SHARE) == null) {
            Log.d(TAG, "Screen share not published")
            return
        }
        try {
            room.localParticipant.setScreenShareEnabled(false)
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Mute microphone (keep track published, no audio) */
    suspend fun muteMic() {
        val room = room ?: return
        try {
            room.localParticipant.setMicrophoneEnabled(false)
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Unmute microphone */
    suspend fun unmuteMic() {
        val room = room ?: return
        try {
            room.localParticipant.setMicrophoneEnabled(true)
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Mute webcam (keep track published, no video) */
    suspend fun muteWebcam() {
        val room = room ?: return
        try {
            room.localParticipant.setCameraEnabled(false)
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Unmute webcam */
    suspend fun unmuteWebcam() {
        val room = room ?: return
        try {
            room.localParticipant.setCameraEnabled(true)
        } catch (e: Exception) {
            callbacks?.onError(e)
            throw e
        }
    }

    /** Disconnect from LiveKit and release all media */
    suspend fun disconnect() {
        room?.disconnect()
        room = null
    }

    /** Fully release the room and resources */
    suspend fun release() {
        room?.release()
        room = null
    }

    fun destroy() {
        scope.cancel()
        room?.release()
        room = null
    }
}